### PR TITLE
Rework ptr-to-ref conversion suggestion for method calls

### LINKED ISSUE #83695: Suggestion for as_mut_ref method on *mut T is <*const T>::as_ref rather than <*mut T>::as_mut
Given the [following code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f15b53bbca5aa0b5125737c87928e17a):

```rust
fn main() {
    let p: *mut () = std::ptr::null_mut();
    p.as_mut_ref();
}
```
(I don't know why to this day, I still type `as_mut_ret`... it's beyond me...)

The current output is:

```
error[E0599]: no method named `as_mut_ref` found for raw pointer `*mut ()` in the current scope
 --> src/main.rs:3:7
  |
3 |     p.as_mut_ref();
  |       ^^^^^^^^^^ method not found in `*mut ()`
  |
  = note: try using `<*const T>::as_ref()` to get a reference to the type behind the pointer: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
  = note: using `<*const T>::as_ref()` on a pointer which is unaligned or points to invalid or uninitialized memory is undefined behavior
```

The note is misleading in two ways:
- it's suggesting a method on `*const T`, when it should be a method on `*mut T`.
- it's suggesting `as_ref` while it should be `as_mut`.

### PR DESCRIPTION
**Pull request by kadiwa4**
If we have a value `z` of type `*const u8` and try to call `z.to_string()`, the upstream compiler will show you a note suggesting to call `<*const u8>::as_ref` first.

This PR extends that:
- The note will only be shown when the method would exist on the corresponding reference type
- It can now suggest any of `<*const u8>::as_ref`, `<*mut u8>::as_ref` and `<*mut u8>::as_mut`, depending on what the method needs.

I didn't introduce a `help` message because that's not a good idea with `unsafe` functions (and you'd also need to unwrap the `Option<&_>` somehow).
People should check the safety requirements.

For the simplest case
```rust
fn main() {
    let x = 8u8;
    let z: *const u8 = &x;
    // issue #21596
    println!("{}", z.to_string()); //~ ERROR E0599
}
```
the output changes like this:
```diff
 error[E0599]: `*const u8` doesn't implement `std::fmt::Display`
   --> $DIR/suggest-convert-ptr-to-ref.rs:5:22
    |
 LL |     println!("{}", z.to_string());
    |                      ^^^^^^^^^ `*const u8` cannot be formatted with the default formatter
    |
-   = note: try using `<*const T>::as_ref()` to get a reference to the type behind the pointer: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
-   = note: using `<*const T>::as_ref()` on a pointer which is unaligned or points to invalid or uninitialized memory is undefined behavior
+note: the method `to_string` exists on the type `&u8`
+  --> $SRC_DIR/alloc/src/string.rs:LL:COL
+   = note: you might want to use the unsafe method `<*const T>::as_ref` to get an optional reference to the value behind the pointer
+   = note: read the documentation for `<*const T>::as_ref` and ensure you satisfy its safety preconditions before calling it to avoid undefined behavior: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
    = note: the following trait bounds were not satisfied:
            `*const u8: std::fmt::Display`
            which is required by `*const u8: ToString`
```

I removed the separate note about the safety requirements because it was incomplete and the linked doc page already has the information you need.

Fixes #83695, but that's more of a side effect. The upstream compiler already suggests the right method name here.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -530,16 +530,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-            err.note(
-                "using `<*const T>::as_ref()` on a pointer which is unaligned or points \
-                 to invalid or uninitialized memory is undefined behavior",
+
+        // on pointers, check if the method would exist on a reference
+        if let SelfSource::MethodCall(rcvr_expr) = source
+            && let ty::RawPtr(ty, ptr_mutbl) = *rcvr_ty.kind()
+            && let Ok(pick) = self.lookup_probe_for_diagnostic(
+                item_name,
+                Ty::new_ref(tcx, ty::Region::new_error_misc(tcx), ty, ptr_mutbl),
```
Comment (kadiwa4) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:539]: I don't know if this lifetime is OK.

**Comment by kadiwa4** [comment]
r? compiler

**Comment by estebank** [comment]
> I removed the separate note about the safety requirements because it was incomplete and the linked doc page already has the information you need.

TBH, I would change it to be

```
+note: the method `to_string` exists on the type `&u8`
+  --> $SRC_DIR/alloc/src/string.rs:LL:COL
+   = note: you might want to use the unsafe method `<*const T>::as_ref` to get an optional reference to the value behind the pointer
+   = warning: using `<*const T>::as_ref()` on a pointer which is unaligned or points to invalid or uninitialized memory is undefined behavior: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
```

or

```
+   = warning: read the documentation for <*const T>::as_ref() and ensure you satisfy its safety preconditions before calling it to avoid undefined behavior: https://doc.rust-lang.org/std/primitive.pointer.html#method.as_ref
```

In general, we can rely on links to have information for things that are nice to know but nor absolutely required. Most people barely read the errors, even fewer read the linked docs.

If we reliably had `<blink>` on the terminal, I'd be tempted to use it for this :stuck_out_tongue:

File: src/doc/book
Comment (estebank) [inline src/doc/book:1]: In this commit you're also modifying a bunch of submodules, which I believe wasn't intentional.

Comment (kadiwa4) [reply src/doc/book:1]: oh, thanks for pointing it out

**Pull request merged**